client.go

```diff
@@ -1,148 +1,192 @@
+        }
 
-		c.rateLoggers[m.Topic].Log("Client frequency", log.Fields{
-			"topic": m.Topic,
-			"from":  m.From,
-		})
+        c.rateLoggers[msg.Topic].Log("Client frequency", log.Fields{
+            "topic": msg.Topic,
+            "from":  msg.From,
+        })
 
-		// Ensure message is coming from correct client.
-		if m.From != c.name || m.PrivateKey != c.privateKey {
-			continue
-		}
-		// Ensure messages are sent in order.
-		if m.Stamp < lastTime {
-			continue
-		}
-		lastTime = m.Stamp
-		c.msgType = m.Type
-		splitTopic := strings.Split(m.Topic, "/")
-        topicName := splitTopic[len(splitTopic)-1]
-        clientName := splitTopic[1]
-		if topicName == DescriptionTopic {
-			var d description
-			json.Unmarshal(m.Msg, &d)
-			if sender, ok := c.h.clientMap[c.privateKey][clientName]; ok {
-				sender.description = d.Data
-			}
-		}
-		for name := range c.h.clientMap[c.privateKey] {
-			if strings.HasPrefix(name, LeafletPrefix) {
-				m.To = append(m.To, name)
-			}
-		}
-		list := make([]string, 0)
-		for _, to := range m.To {
-			if sender, ok := c.h.clientMap[c.privateKey][to]; ok {
-				exists := false
-				for _, check := range list {
-					if check == to {
-						exists = true
-						break
-					}
-				}
-				if !exists {
-					list = append(list, to)
-					snd := SendPacket{addr: sender.addr, data: append([]byte{}, msg...)}
-					select {
-					case c.h.sendChannel <- snd:
-					default:
-					}
-				}
-			} else { // Regex
-				for name, sender := range c.h.clientMap[c.privateKey] {
-					if name != m.From {
-						match, _ := regexp.MatchString(to, name)
-						if match {
-							exists := false
-							for _, check := range list {
-								if check == name {
-									exists = true
-									break
-								}
-							}
-							if !exists {
-								list = append(list, name)
-								snd := SendPacket{addr: sender.addr, data: append([]byte{}, msg...)}
-								select {
-								case c.h.sendChannel <- snd:
-								default:
-								}
-							}
-						}
-					}
-				}
-			}
-		}
-	}
+        // Ensure message is coming from correct client.
+        if msg.From != c.name || msg.PrivateKey != c.privateKey {
+            continue
+        }
+        // Ensure messages are sent in order.
+        if msg.Stamp < lastTime {
+            continue
+        }
+        lastTime = msg.Stamp
+        c.msgType = msg.Type
+        topicName, clientName := processTopicString(msg.Topic)
+        if topicName == DescriptionTopic {
+            err = c.ProcessDescription(&msg, clientName)
+            if err != nil {
+                continue
+            }
+        }
+        for name := range c.hub.clientMap[c.privateKey] {
+            if strings.HasPrefix(name, LeafletPrefix) {
+                msg.To = append(msg.To, name)
+            }
+        }
+        c.SendMessage(&msg, &packet)
+    }
+}
+
+// UnpackMessage converts raw packet data into a Message struct
+func (c *Client) UnpackMessage(packet *[]byte) (Message, error) {
```

Comment:
Nice, this is a lot easier to read with this logic moved into a separate function!